### ISSUE TITLE
constexpr check if type is loggable

### ISSUE THREAD

**Issue by andrewkcorcoran** [morganstanley/binlog#145]
Is it possible to check (in a constexpr context) if a given type is serializable by binlog? Motivating example
```cpp
template<typename T>
void Log(T && t)
{
   if constexpr(binlog::is_serializable<T>)
   {
      BINLOG_INFO("{}", t);
   }
   else
   {
      BINLOG_INFO("Non serializable type t: {}", boost::typeindex::type_id<T>().pretty_name());
   }
}
```

**Comment by erenon** [comment]
It is possible using `mserialize::detail::is_serializable`, e.g:

https://github.com/morganstanley/binlog/blob/main/test/integration/LoggingBoostTypes.cpp#L63

**Comment by andrewkcorcoran** [comment]
```bash
include/mserialize/detail/Tag.hpp:31:5: error: static_assert failed due to requirement 'always_false<SNIP>::value' "T has no associated tag"
```
`mserialize::detail::is_serializable<TEvent>::value` is returning true but then the above static assert is failing? Does is_serializable not check for tags, is there a second constexpr check that needs to be done to guarantee a type will pass into binlog?

**Comment by erenon** [comment]
(Sorry, I was giving the literal answer about checking if the type is serializable, though I should have suspected you are after checking if it is loggable. As you figured out, both is_serializable and has_tag are required.)

Can you give an example of a type `T` that satisfies has_tag, but mserialize::tag<T> still asserts?

**Comment by andrewkcorcoran** [comment]
Looks to be when an adapted struct has a serialisable type which has not been adapted.

Minimal example below.

```cpp
struct Nested{};

template<typename TComposed>
struct Composed
{
  TComposed composed;
};
BINLOG_ADAPT_TEMPLATE((typename TComposed), (Composed<TComposed>), composed)

template<typename T>
void Log(const T & t)
{
  if constexpr(mserialize::detail::is_serializable<T>::value && mserialize::detail::has_tag<T>::value)
  {
	  BINLOG_INFO("{}", t);
  }
}

int main()
{
  Log(Composed<Nested>{});
}
```

**Comment by erenon** [comment]
With `BINLOG_ADAPT_STRUCT` instead of `BINLOG_ADAPT_TEMPLATE` the problem becomes more obvious: the fields must be also adapted - this is a (documented) requirement of the macro, but because of the late evaluation of the template, it only blows up on usage (compared to `BINLOG_ADAPT_STRUCT`, that fails immediately).

A fix I can think of: add static_asserts to `BINLOG_ADAPT_TEMPLATE`, and make this requirement explicit. Do you have a different use-case that wouldn't be covered by this?

**Comment by andrewkcorcoran** [comment]
How would a static assert in `BINLOG_ADAPT_TEMPLATE` work in the above example? `BINLOG_ADAPT_TEMPLATE` doesn't know what the type of the template is going to be, so therefore it can't assert on the template type having being adapted.

The general requirement here is a way to write a templated log method which will log an event as normal if binlog supports that event or otherwise fall back to just logging the type name or some other generic approach. I think the only way for this to work is if `is_serializable` and `has_tag` traverse the full recursive tree of types, the same way the log macros do.

**Comment by erenon** [comment]
Ok, a static_assert wouldn't help. I either need to make the generated specialization conditionally-constructible (I don't know if that's possible) or change the definition of has_tag, to check the existence of a nested type (for example), that can be recursively produced.

**Comment by erenon** [comment]
> How would a static assert in BINLOG_ADAPT_TEMPLATE work in the above example? BINLOG_ADAPT_TEMPLATE doesn't know what the type of the template is going to be, so therefore it can't assert on the template type having being adapted.

The static_assert can depend on the type of the template parameter, and fire only when the template is instantiated. Do you use `BINLOG_ADAPT_TEMPLATE` in a way that the fields are may or may not be adapted?

**Comment by andrewkcorcoran** [comment]
Yes there's no guarantee that the composed type is adaptable. In our use case this, we receive events, wrap them in the template class and then a later stage tries to log them. We don't directly control the events we receive so some of them may have been adapted, some not.

**Comment by erenon** [comment]
Please see if this works for you: https://github.com/morganstanley/binlog/pull/146

**Comment by andrewkcorcoran** [comment]
Still seems to be failing on the log line if you add the below to the test case in PR.

```cpp
int main()
{
    auto e = Nest<NotAdapted>{};
    if constexpr(mserialize::detail::has_tag<Nest<NotAdapted>>::value && mserialize::detail::is_serializable<Nest<NotAdapted>>::value)
    {
        BINLOG_INFO("{}", e);
    }
    return 0;
}
```

**Comment by erenon** [comment]
According to cppreference, the condition of the `if constexpr` must depend on a template argument, otherwise the discarded statement is fully checked. 

> Outside a template, a discarded statement is fully checked. if constexpr is not a substitute for the [#if](https://en.cppreference.com/w/cpp/preprocessor/conditional) preprocessing directive:

https://en.cppreference.com/w/cpp/language/if

I tested with the original example `Log`, and it was working.

**Comment by erenon** [comment]
Re variant: New testcase and fix added to #146

**Comment by andrewkcorcoran** [comment]
Looks good!
